vault/resource_kmip_secret_backend.go

```diff
@@ -0,0 +1,222 @@
+			},
+			"default_tls_client_key_bits": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Computed:    true,
+				Description: "Client certificate key bits, valid values depend on key type",
+			},
+			"default_tls_client_ttl": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Computed:    true,
+				Description: "Client certificate TTL in seconds",
+			},
+		},
+	}
+}
+
+func kmipSecretBackendCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Get("path").(string)
+	description := d.Get("description").(string)
+	defaultTLSClientTTL := d.Get("default_tls_client_ttl").(int)
+
+	log.Printf("[DEBUG] Mounting KMIP backend at %q", path)
+	if err := client.Sys().Mount(path, &api.MountInput{
+		Type:        "kmip",
+		Description: description,
+		Config: api.MountConfigInput{
+			DefaultLeaseTTL: fmt.Sprintf("%d", defaultTLSClientTTL),
+		},
+	}); err != nil {
+		return fmt.Errorf("error mounting to %q: %s", path, err)
+	}
+
+	log.Printf("[DEBUG] Mounted KMIP backend at %q", path)
+	d.SetId(path)
+
+	return kmipSecretBackendUpdate(d, meta)
+}
+
+func kmipSecretBackendUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Id()
+	tune := api.MountConfigInput{}
+
+	if d.HasChange("default_tls_client_ttl") || d.HasChange("description") {
+		tune.DefaultLeaseTTL = fmt.Sprintf("%ds", d.Get("default_tls_client_ttl"))
+		description := d.Get("description").(string)
+		tune.Description = &description
+
+		log.Printf("[DEBUG] Updating mount for %q", path)
+		err := client.Sys().TuneMount(path, tune)
+		if err != nil {
+			return fmt.Errorf("error updating mount for %q, err=%w", path, err)
+		}
+		log.Printf("[DEBUG] Updated mount for %q", path)
+	}
+
+	data := map[string]interface{}{}
+	configPath := fmt.Sprintf("%s/config", path)
+	log.Printf("[DEBUG] Updating %q", configPath)
+
+	for _, k := range kmipAPIFields {
+		if d.HasChange(k) {
+			if v, ok := d.GetOk(k); ok {
+				switch v.(type) {
+				case *schema.Set:
+					data[k] = util.TerraformSetToStringArray(v)
+				default:
+					data[k] = v
+				}
+			}
+		}
+	}
+
+	if _, err := client.Logical().Write(configPath, data); err != nil {
+		return fmt.Errorf("error updating KMIP config %q: %s", configPath, err)
+	}
+
+	log.Printf("[DEBUG] Updated %q", configPath)
+
+	return kmipSecretBackendRead(d, meta)
+}
+
+func kmipSecretBackendRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	path := d.Id()
+	log.Printf("[DEBUG] Reading KMIP config at %s/config", path)
+	resp, err := client.Logical().Read(path + "/config")
+	if err != nil {
+		return fmt.Errorf("error reading KMIP config at %q/config: %s", path, err)
+	}
+
+	if resp == nil {
+		log.Printf("[WARN] KMIP config not found, removing from state")
+		d.SetId("")
+
+		return nil
+	}
+
+	for _, k := range kmipAPIFields {
+		if err := d.Set(k, resp.Data[k]); err != nil {
```

Comment:
What if `k` is not in `resp.Data`?

Reply:
Then `resp.Data[k]` would result in ` nil`, which is what we want that value to be set to in that case, right?

Reply:
I am more thinking of the case where the target Vault version does provide `k` in its response?

Reply:
I see, would something along this lines be better then:
```
        for _, k := range kmipAPIFields {
		if v, ok := resp.Data[k]; ok {
			if err := d.Set(k, v); err != nil {
				return fmt.Errorf("error setting state key %q on KMIP config, err=%w", k, err)
			}
		}
	}

```

Reply:
Yeah, I think as you mentioned above, we should be okay with a `nil` value for any schema Type. Our schema must always be the superset of all supported Vault versions.